### PR TITLE
Acroform checkbox and radio controls

### PR DESCRIPTION
**Pull request by benweet**
Added fake checkboxes and radio buttons with a pinch of css.
Tested against http://www.cic.gc.ca/english/passport/forms/pdf/pptc153.pdf on Chrome and Firefox:

screen shot 2016-09-22 at 16 19 06

### CODE REVIEW

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Why are we toggling the class name here? I would expect that we set the `checked` *attribute* and style that with CSS (using the `input:checked` selector) so we don't have to use classes here.

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Indent with two spaces and remove the newline below.

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: *radio button* widget

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Add a space after `?`. I find it a bit hard to understand what this line does exactly. Can we break it up a bit?

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Change this to `for (var i = 0, ii = otherKids.length; i < ii; i++) {` to cache the length.

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: My previous comment about the class name applies here too. It's a bit confusing why we do that.

File: web/annotation_layer_builder.css
Comment (timvandermeij) [inline web/annotation_layer_builder.css]: Is this styling also how it looks in Adobe Reader? If not, I'd prefer we keep it the same as Adobe Reader for now as that's what most users will expect.

**Review by timvandermeij — CHANGES_REQUESTED** [review]
This looks really good in general. I'm happy that you made this work with the new framework we're developing!

Aside from these review comments, I'd like to ask you to write unit tests for this. You can find these in `test/unit/annotation_layer_spec.js`. When this is done, I may be able to create a reference test for this patch.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Why is this patch using `<div>`s and a lot of complicated CSS instead of just standard HTML form elements (with appropriate styling)?
One obvious downside to the current approach is that the elements added in the patch isn't accessible as the other ones. E.g. note how it's not possible to <kbd>tab</kbd> to them, which using proper HTML form elements would fix.

Also, before we start building upon it, we really ought to go through and attempt to clean-up/refactor this code: https://github.com/mozilla/pdf.js/blob/6c263c19946af23b723f148d9f05118971e18b36/src/core/annotation.js#L640-L670

Finally, this patch is currently missing both unit and reference tests!

File: web/annotation_layer_builder.css
Comment (benweet) [inline web/annotation_layer_builder.css]: That's the color used by you guys in `.textWidgetAnnotation`. It looks pretty much like Adobe Reader.

File: src/display/annotation_layer.js
Comment (benweet) [inline src/display/annotation_layer.js]: Divs are easier to style and resize well. Would hidden input be acceptable? Something like http://www.inserthtml.com/demos/css/radio-buttons/

**Comment by benweet** [comment]
@timvandermeij I'll take your remarks into account and add unit tests.

**Comment by timvandermeij** [comment]
I have been looking into this a bit more and I now realize why you chose `div`s. I forgot that it's not really possible to style `input` elements other than those with `type="text"`. We do need `input` elements, however, both for accessibility as well as for easily reading the selected options, so I think your proposal for using hidden inputs is the only possible way to both make this look good and work like regular `input` elements, so I'm okay with that.

**Comment by Snuffleupagus** [comment]
> [...] so I think your proposal for using hidden inputs is the only possible way to both make this look good and work like regular input elements, so I'm okay with that.

My suspicion is that using hidden elements will unfortunately do nothing to alleviate the keyboard accessibility problems!
What I'd suggest is to use regular HTML form elements for now (e.g. checkbox, radiobutton, etc.), since they should be easy to implement and can be improved later on. (Given that there's still a number of major roadblocks to enabling forms by default, e.g. printing/saving the entered data and `Appearance` stream support.)
However, one possible way forward instead of using HTML form elements (e.g. checkbox, radiobutton, etc) *could* perhaps be to utilize `<button>` elements instead, since they are possible to style as you want and should still give proper accessibility.

**Comment by timvandermeij** [comment]
Right, it's a bit unfortunate, but the current state of the web platform won't really let us do what we want. Using `<button>` elements might help, along with some "checked" state in the form of a class as is done now.

@benweet Could you check if using `<button>` elements might help, or if you can improve accessibility (tab the hidden input fields) somehow? It could help determine what the best way to get is here. If those do not work, I also suggest to just use regular checkboxes and radio buttons without styles.

**Comment by benweet** [comment]
Here is a new commits with hidden inputs. Keyboard accessibility works fine, though I didn't manage to have the highlighting on focus.
@timvandermeij I've taken your comments into account but I still have to add tests.

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: The background colour *needs* to be transparent, please undo this change.

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: Should be the same colour as above, i.e. `background-color: rgba(0, 54, 255, 0.13);`

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: The condition doesn't actually agree with the comment, so change it to:
`if (isName(fieldValue)) {` instead (and remove the comment since it's then superfluous).

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: There seems to be some overlap between these rules and the ones above, could that be simplified somewhat?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
> [...] though I didn't manage to have the highlighting on focus.

I'm afraid that this is something that will need to be addressed somehow, if this current solution is to work, since it's not particularily useful to be able to <kbd>tab</kbd> to the element if the user cannot see that it has focus.

File: web/annotation_layer_builder.css
Comment (benweet) [inline web/annotation_layer_builder.css]: My mistake, sorry about that.

File: web/annotation_layer_builder.css
Comment (benweet) [inline web/annotation_layer_builder.css]: When the pdf is pre-filled, the checked box is drawn in the canvas. I changed the color to be opaque so that the checked box is hidden behind the annotation.

Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: I cannot reproduce this, please provide a PDF file that demonstrates this problem!
Without having seen the file, it sounds like it may be an issue related to `Appearance` streams support in conjunction with forms, which is a somewhat difficult issue that we'll have to somehow fix before we can enable forms by default.

However, simply hiding the background in this way isn't going to be a working solution, so please change this for now.

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js]: Is this warning really necessary though, since it's going to be printed in the console regardless if forms are enabled or not?
Given that forms support is still some ways from being implemented, is it necessary to bother users about this?

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: The [PDF specification](http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdf#G11.1952008) specifically mentions that the `RADIO` flag cannot be set at the same time as the `PUSHBUTTON` flag, but that is missing from the code here.

File: web/annotation_layer_builder.css
Comment (benweet) [inline web/annotation_layer_builder.css]: Here is a sample pdf: [pptc153filled.pdf](https://github.com/mozilla/pdf.js/files/491352/pptc153filled.pdf)

File: src/core/annotation.js
Comment (benweet) [inline src/core/annotation.js]: Do we really want to check that the document complies with the PDF spec? What does it bring?

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: As I suspected, that's an issue with `Apperance` streams, so please change this as I originally suggested since the proper solution lie elsewhere.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: Well, PDF.js is a *PDF viewer* after all, so we most certainly want to try and follow the specification!
If experience has thought us anything, it's that PDF files found in the wild breaks the specification in all sorts of "interesting" ways, which often leads to bugs. So of course we always want try to ensure that data present in a PDF file is valid!

Comment (benweet) [inline src/core/annotation.js]: Right. What do you suggest?

Note that I've already put a bit of logic in `annotation_layer.js` that does some kind of compliancy check:

```
if (!parameters.data.pushbutton) {
    if (parameters.data.radio) {
       return new RadioButtonWidgetAnnotationElement(parameters);
    } else {
       return new CheckboxWidgetAnnotationElement(parameters);
    }
}
```

File: web/annotation_layer_builder.css
Comment (benweet) [inline web/annotation_layer_builder.css]: Can we just use an opaque color for the time being and keep this as a future enhancement?

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: The main validation should happen in the `core` layer, so wouldn't e.g. this be enough:
```javascript
this.data.radio = this.hasFieldFlag(AnnotationFieldFlag.RADIO) &&
                  !this.hasFieldFlag(AnnotationFieldFlag.PUSHBUTTON);
```

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css]: Why not try to fix this properly instead, by adding `getOperatorList` overrides to the new class (in `src/core/annotation.js`)?
Similar to the existing `TextWidgetAnnotation`, but perhaps a bit simpler for now:
```javascript
Util.inherit(ButtonWidgetAnnotation, WidgetAnnotation, {
  getOperatorList:
      function ButtonWidgetAnnotation_getOperatorList(evaluator, task,
                                                      renderForms) {
    var operatorList = new OperatorList();

    // Do not render form elements on the canvas when interactive forms are
    // enabled. The display layer is responsible for rendering them instead.
    if (renderForms) {
      return Promise.resolve(operatorList);
    }

    if (this.appearance) {
      return Annotation.prototype.getOperatorList.call(this, evaluator, task,
                                                       renderForms);
    }
    return Promise.resolve(operatorList);
  }
```

Comment (benweet) [inline web/annotation_layer_builder.css]: Works pretty good! I'll merge this tomorrow. Thanks

**Comment by benweet** [comment]
Added unit tests and CSS outline on focus.

**Comment by timvandermeij** [comment]
Let's remove the `noToggleToOff` code if we're not implementing it. Then, please squash the commits into one commit so we can do a new review round. If you're not familiar with squashing commits, here is how to do it easily: https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

**Comment by benweet** [comment]
Done.
Cheers!

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: You can use `this.data.fieldValue` here after the other PR is merged, but that's just a note for now.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: What happens if it's not a `Name`? It happens very often that PDF files do not adhere to the specification entirely, so if they do not provide a `Name` here, then `this.data.fieldValue` will be used as-is after the other PR is merged. Should we make it empty in such a case?

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Perhaps not really. I'm also fine with a comment such as `// Push buttons are not implemented yet` so that we at least know in the code that something is still missing here.

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Let's add an empty line before this line to make this a bit more readable as a new block of code.

File: src/display/annotation_layer.js
Comment (timvandermeij) [inline src/display/annotation_layer.js]: Same here, let's add a newline above this line for readability.

File: test/unit/annotation_layer_spec.js
Comment (timvandermeij) [inline test/unit/annotation_layer_spec.js]: For all occurrences in this file: move this to the previous line. We only use this if it doesn't fit on one line anymore.

File: test/unit/annotation_layer_spec.js
Comment (timvandermeij) [inline test/unit/annotation_layer_spec.js]: For all occurrences in this file: remove two spaces from the indentation here. We indent everything with two spaces.

**Review by timvandermeij — CHANGES_REQUESTED** [review]
This patch looks much better! Let's wait until the other PR is merged and then address these comments. Thanks!

**Comment by timvandermeij** [comment]
Everything that blocked this PR has landed, so it's now good to rebase and fix up, @benweet. Thanks.

**Comment by benweet** [comment]
Fixed. Cheers!

**Comment by timvandermeij** [comment]
Closing as replaced by #7898, which contains the commit from this PR and follow-up fixes so we can land it.

**Pull request closed without merging**
